### PR TITLE
More logging for Supermatter activations & cleaning?.

### PR DESCRIPTION
**Pull request by nevimer**
## About The Pull Request

This pull request adds logging and flags the start of the supermatter from environmental starts. It also makes the log generation  and started flag flipping a proc. This means that atmospheric starts are now logged (co2, o2, trit, etc), and so is touching/running into the SM. 

[Dreamseeker 9Gkzvyfwdp-1.webm](https://user-images.githubusercontent.com/77420409/195494711-d9295b23-3ec3-447e-b255-fd14454bab5c.webm)

## Why It's Good For The Game

Just a tad more log clarity when you go to investigate. Allows accidental/intentional/grief starts to be diagnosed easier.

### CODE REVIEW

**Comment by vincentiusvin** [comment]
Would you mind renaming the proc to emphasize that it just does logging? I freaked out a bit when i first read it because I thought it was going to change how the rock works.

Something like log_activation or whatever else that floats your boat.

**Comment by nevimer** [comment]
> Would you mind renaming the proc to emphasize that it just does logging? I freaked out a bit when i first read it because I thought it was going to change how the rock works.
> 
> Something like log_activation or whatever else that floats your boat.

yea sure

File: code/modules/power/supermatter/supermatter.dm [hunk 1]
```diff
@@ -264,6 +264,8 @@ GLOBAL_DATUM(main_supermatter_engine, /obj/machinery/power/supermatter_crystal)
 	internal_energy_factors = calculate_internal_energy()
 	zap_factors = calculate_zap_multiplier()
 	if(internal_energy && (last_power_zap + 4 SECONDS - (internal_energy * 0.001)) < world.time)
+		if(!has_been_powered)
+			log_activation(source = "the environment at [get_area_name(local_turf)]")
```
Comment (LemonInTheDark) [inline code/modules/power/supermatter/supermatter.dm:268]: won't this lead to us trying to get the key_name of this string?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/modules/power/supermatter/supermatter.dm (hunk 1 above)
Comment (nevimer) [reply code/modules/power/supermatter/supermatter.dm:268]: It will work text-wise but you just can't jump to it.

Comment (LemonInTheDark) [reply code/modules/power/supermatter/supermatter.dm:268]: hm

**Comment by LemonInTheDark** [comment]
@nevimer I recall discussing this with you on discord, did you ever do anythin about that?

**Comment by nevimer** [comment]
> @nevimer I recall discussing this with you on discord, did you ever do anythin about that?

I forgot about it TBH but I'll finish this up if you reopen it

File: code/modules/power/supermatter/supermatter_hit_procs.dm [hunk 2]
```diff
@@ -44,6 +32,22 @@
...
+	var/fired_from_str = cause ? " with [cause]" : ""
+	investigate_log(
+		source \
+			? "has been powered for the first time by [key_name(source)][fired_from_str]." \
+			: "has been powered for the first time.",
+		INVESTIGATE_ENGINE
+	)
+	message_admins(
+		source \
+			? "[src] [ADMIN_JMP(src)] has been powered for the first time by [cause ? ADMIN_FULLMONTY(source) + (fired_from_str) : (source ? source : "environmental factors")]." \
```
Comment (LemonInTheDark) [inline code/modules/power/supermatter/supermatter_hit_procs.dm:45]: why the second source ? : here?

Comment (nevimer) [reply code/modules/power/supermatter/supermatter_hit_procs.dm:45]: if a source is not passed or non-atomic it falls back. so theres always something if any combo happens

Comment (LemonInTheDark) [reply code/modules/power/supermatter/supermatter_hit_procs.dm:45]: yeah but we can't get to this branch of logic unless there's a source, right?

Comment (nevimer) [reply code/modules/power/supermatter/supermatter_hit_procs.dm:45]: oh right maybe I intended it to be a cause

**Comment by Mothblocks** [comment]
Stale

**Comment by nevimer** [comment]
> 



> Stale

not stale, pretty much done

File: code/modules/power/supermatter/supermatter_hit_procs.dm (hunk 2 above)
Comment (nevimer) [reply code/modules/power/supermatter/supermatter_hit_procs.dm:45]: ima bap this real fast, in any case the messages that are spat out are human readable and make sense

**Pull request merged**